File: TS/Basic/Collections/seqs.ts

export function seqsExample() {
  const a = [1, 2, 3]
  let b = a.map(x => x * x)
  const c = a.filter(x => x % 2 === 0)


  // iterate elem and index
  a.forEach((v, i) => {
    b[i] = v * v
  })

  for (let i = 4; i <= 10; i++) {
    b.push(i * i)    
  }

  console.log("b = ", b);

  // Sum of lists
  const sumList = a.concat(b) // or 
  const sumList2 = [...a, ...b]
  console.log(sumList)

  // remove item
  a.splice(0, 1)
  b = [a[0], ...b]
  
  // there no immutability
}

File: Nim/old/Collections/seqs.nim
import std/[sequtils]

var
  a = @[1, 2, 3]
  b = a.mapIt(it * it)
  c = a.filterIt(it mod 2 == 0)

# iterate elem and index
for i, v in a:
  b[i] = v*v

for i in 4..10:
  b.add(i * i)

echo "b = ", b

# Sum of lists
let sumList = a & b
echo sumList

# remove item
b.delete(0)  # takes O(n) time
b = a[0] & b  # Same as original b


# Immutability

let d = @[1, 2, 3]
d.add(4) # ERROR

var e = @[1, 2, 3]
e.add(4)

proc doSomething(mySeq: seq[int]) =
  mySeq[0] = 2  # this is a compile-time error
var testSeq = @[1, 2, 3]
doSomething(testSeq)

proc foo(mySeq: var seq[int]) =
  mySeq[9] = 999

var thisSeq = newSeq[int](10)
foo(thisSeq)

assert thisSeq[9] == 999

# send by value, modify only copy

proc doSomething2(mySeq: seq[int]) =
  var varMySeq = mySeq  # copy the seq
  varMySeq[0] = 999
  assert varMySeq[0] == 999
var testSeq2 = @[1, 2, 3]
doSomething(testSeq)
assert testSeq[0] == 1